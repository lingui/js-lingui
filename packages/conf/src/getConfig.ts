import fs from "fs"
import { LinguiConfigNormalized } from "./types"
import { lilconfigSync, LoaderSync } from "lilconfig"
import yaml from "yaml"
import path from "path"
import { makeConfig } from "./makeConfig"
import type { JITIOptions } from "jiti/dist/types"
import pico from "picocolors"

function configExists(path: string) {
  return path && fs.existsSync(path)
}

function JitiLoader(): LoaderSync {
  return (filepath) => {
    const opts: JITIOptions = {
      interopDefault: true,
    }
    const jiti = require("jiti")(__filename, opts)
    return jiti(filepath)
  }
}

function YamlLoader(): LoaderSync {
  return (_, content) => {
    console.warn(
      "YAML config support is deprecated and will be removed in future versions."
    )
    return yaml.parse(content)
  }
}

const moduleName = "lingui"

const configExplorer = lilconfigSync(moduleName, {
  searchPlaces: [
    `.${moduleName}rc`,
    `.${moduleName}rc.json`,
    `.${moduleName}rc.js`,
    `.${moduleName}rc.ts`,
    `.${moduleName}rc.yaml`,
    `.${moduleName}rc.yml`,
    `${moduleName}.config.js`,
    `${moduleName}.config.cjs`,
    `${moduleName}.config.ts`,
    `${moduleName}.config.mjs`,
    "package.json",
  ],
  loaders: {
    ".js": JitiLoader(),
    ".ts": JitiLoader(),
    ".mjs": JitiLoader(),
    ".yaml": YamlLoader(),
    ".yml": YamlLoader(),
  },
})

export function getConfig({
  cwd,
  configPath,
  skipValidation = false,
}: {
  cwd?: string
  configPath?: string
  skipValidation?: boolean
} = {}): LinguiConfigNormalized {
  const defaultRootDir = cwd || process.cwd()

  configPath = configPath || process.env.LINGUI_CONFIG

  let result

  try {
    result = configExists(configPath)
      ? configExplorer.load(configPath)
      : configExplorer.search(defaultRootDir)
  } catch (error) {
    // If lilconfig throws an error (e.g., directory doesn't exist), treat it as no config found
    result = null
  }

  if (!result) {
    console.error("Lingui was unable to find a config!\n")
    console.error(
      `Create ${pico.bold(
        "'lingui.config.js'"
      )} file with LinguiJS configuration in root of your project (next to package.json). See ${pico.underline(
        "https://lingui.dev/ref/conf"
      )}`
    )

    // gracefully stop further executing
    throw new Error("No Lingui config found")
  }

  const userConfig = result ? result.config : {}

  return makeConfig(
    {
      rootDir: result ? path.dirname(result.filepath) : defaultRootDir,
      ...userConfig,
    },
    { skipValidation }
  )
}
