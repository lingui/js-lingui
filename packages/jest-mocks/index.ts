export function getConsoleMockCalls({ mock }: jest.MockInstance<any, any>) {
  if (!mock.calls.length) return
  return mock.calls.map((call) => call[0]).join("\n")
}

export function mockConsole(
  testCase: (console: jest.Mocked<Console>) => any,
  mock = {}
) {
  function restoreConsole() {
    global.console = originalConsole
  }

  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  }

  global.console = {
    ...defaults,
    ...mock,
  } as any

  let result
  try {
    result = testCase(global.console as jest.Mocked<Console>)
  } catch (e) {
    restoreConsole()
    throw e
  }

  if (result && typeof result.then === "function") {
    return result.then(restoreConsole).catch((e) => {
      restoreConsole()
      throw e
    })
  } else {
    restoreConsole()
    return result
  }
}

export function mockEnv(env, testCase) {
  const oldEnv = process.env.NODE_ENV
  process.env.NODE_ENV = env

  try {
    testCase()
  } catch (e) {
    process.env.NODE_ENV = oldEnv
    throw e
  }

  process.env.NODE_ENV = oldEnv
}
