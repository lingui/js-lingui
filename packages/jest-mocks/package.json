{
  "private": true,
  "name": "@lingui/jest-mocks",
  "version": "0.0.0-managed-by-release-script",
  "description": "Mocks for Jest",
  "main": "index.js",
  "author": {
    "name": "Tomáš Ehrlich",
    "email": "tomas.ehrlich@gmail.com"
  },
  "license": "MIT",
  "keywords": [
    "jest",
    "testing",
    "mock"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/lingui/js-lingui.git"
  },
  "bugs": {
    "url": "https://github.com/lingui/js-lingui/issues"
  },
  "engines": {
    "node": ">=6.0"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.js",
    "*.js.flow"
  ]
}
