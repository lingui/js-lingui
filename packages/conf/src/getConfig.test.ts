import path from "path"
import { getConfig } from "./getConfig"
import { mockConsole } from "@lingui/jest-mocks"

describe("getConfig", () => {
  const fixturesPath = path.resolve(__dirname, "fixtures")
  const validPath = path.join(fixturesPath, "valid")

  describe("config file discovery", () => {
    it("should find and load .linguirc config file", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, ".linguirc"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["en-gb"])
        expect(config.sourceLocale).toEqual("en-gb")
      })
    })

    it("should load JavaScript config file", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "lingui.config.js"),
          skipValidation: true,
        })

        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["en", "fr"])
        expect(config.sourceLocale).toEqual("en")
      })
    })

    it("should load TypeScript config file", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "lingui.config.ts"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["en", "it"])
        expect(config.sourceLocale).toEqual("en")
      })
    })

    it("should load CommonJS config file", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "lingui.config.cjs"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["en", "de"])
        expect(config.sourceLocale).toEqual("en")
      })
    })

    it("should load ES module config file", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "lingui.config.mjs"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["en", "es"])
        expect(config.sourceLocale).toEqual("en")
      })
    })

    it("should load YAML config file with deprecation warning", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "lingui.config.yaml"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          "YAML config support is deprecated and will be removed in future versions."
        )
        expect(config.locales).toEqual(["en", "ru"])
        expect(config.sourceLocale).toEqual("en")
      })
    })

    it("should load YML config file with deprecation warning", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "lingui.config.yml"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          "YAML config support is deprecated and will be removed in future versions."
        )
        expect(config.locales).toEqual(["en", "pt"])
        expect(config.sourceLocale).toEqual("en")
      })
    })

    it("should load config from package.json", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "package.json"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["en", "zh"])
        expect(config.sourceLocale).toEqual("en")
      })
    })

    it("should load specific config file with configPath parameter", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "custom.config.js"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["cs", "sk"])
        expect(config.sourceLocale).toEqual("cs")
      })
    })

    it("should load TypeScript custom config file", () => {
      mockConsole((console) => {
        const config = getConfig({
          configPath: path.join(validPath, "custom.config.ts"),
          skipValidation: true,
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config.locales).toEqual(["pl"])
        expect(config.sourceLocale).toEqual("pl")
      })
    })
  })

  describe("error handling", () => {
    it("should throw error if config is not found", () => {
      mockConsole((console) => {
        const exec = () =>
          getConfig({
            cwd: fixturesPath,
          })

        expect(exec).toThrow("No Lingui config found")
        expect(console.error).toBeCalledWith(
          "Lingui was unable to find a config!\n"
        )
        expect(console.error).toBeCalledWith(expect.stringContaining("Create"))
        expect(console.warn).not.toBeCalled()
      })
    })

    it("should throw error if specific config file doesn't exist", () => {
      mockConsole((console) => {
        const exec = () =>
          getConfig({
            configPath: path.join(validPath, "nonexistent.config.js"),
          })

        expect(exec).toThrow("No Lingui config found")
        expect(console.error).toBeCalledWith(
          "Lingui was unable to find a config!\n"
        )
        expect(console.warn).not.toBeCalled()
      })
    })
  })

  describe("environment variable support", () => {
    it("should use LINGUI_CONFIG environment variable", () => {
      const originalEnv = process.env.LINGUI_CONFIG
      process.env.LINGUI_CONFIG = path.join(validPath, "custom.config.js")

      try {
        mockConsole((console) => {
          const config = getConfig({ skipValidation: true })
          expect(console.error).not.toBeCalled()
          expect(console.warn).not.toBeCalled()
          expect(config.locales).toEqual(["cs", "sk"])
          expect(config.sourceLocale).toEqual("cs")
        })
      } finally {
        process.env.LINGUI_CONFIG = originalEnv
      }
    })
  })

  describe("default config and snapshots", () => {
    it("should return default config", () => {
      mockConsole((console) => {
        const config = getConfig({
          cwd: path.resolve(__dirname, path.join("fixtures", "default")),
        })
        expect(console.error).not.toBeCalled()
        expect(console.warn).not.toBeCalled()
        expect(config).toMatchSnapshot()
      })
    })
  })
})
