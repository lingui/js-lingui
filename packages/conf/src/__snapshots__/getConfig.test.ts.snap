// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConfig default config and snapshots should return default config 1`] = `
{
  catalogs: [
    {
      exclude: [
        */node_modules/*,
      ],
      include: [
        .,
      ],
      path: ./locale/{locale}/messages,
    },
  ],
  catalogsMergePath: ,
  compileNamespace: cjs,
  compilerBabelOptions: {
    jsescOption: {
      minimal: true,
    },
    minified: true,
  },
  extractorParserOptions: {
    flow: false,
    tsExperimentalDecorators: false,
  },
  fallbackLocales: {
    en-gb: en,
  },
  format: po,
  formatOptions: {
    lineNumbers: true,
    origins: true,
  },
  locales: [
    en-gb,
  ],
  macro: {
    corePackage: [
      @lingui/macro,
      @lingui/core/macro,
    ],
    jsxPackage: [
      @lingui/macro,
      @lingui/react/macro,
    ],
  },
  orderBy: message,
  pseudoLocale: ,
  rootDir: .,
  runtimeConfigModule: {
    Trans: [
      @lingui/react,
      Trans,
    ],
    i18n: [
      @lingui/core,
      i18n,
    ],
    useLingui: [
      @lingui/react,
      useLingui,
    ],
  },
  service: {
    apiKey: ,
    name: ,
  },
  sourceLocale: en-gb,
}
`;
